onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc

```diff
@@ -0,0 +1,198 @@
+    std::vector<int64_t> pos_id_dim({batch_size, sequence_length});
+    gsl::span<const int64_t> pos_id_dim_span{pos_id_dim};
+    TensorShape pos_id_shape(pos_id_dim_span);
+    Tensor::InitOrtValue(element_type, pos_id_shape, allocator, pos_ids);
+    ORT_RETURN_IF_ERROR(group_query_attention_helper::GeneratePositionIds(seqlens_k->Data<int32_t>(), batch_size,
+                                                                          sequence_length, pos_ids.GetMutable<Tensor>()->MutableData<int64_t>(), context));
+
+    const T* q_input;
+    const T* k_input;
+    T* q_rotary;
+    T* k_rotary;
+    if (packed_qkv) {
+      OrtValue RotaryQKV;
+      element_type = DataTypeImpl::GetType<T>();
+      std::vector<int64_t> qkv_dim({batch_size, num_heads_ + 2 * kv_num_heads_, sequence_length, head_size});
+      gsl::span<const int64_t> qkv_dim_span{qkv_dim};
+      TensorShape qkv_shape(qkv_dim_span);
+      Tensor::InitOrtValue(element_type, qkv_shape, allocator, RotaryQKV);
+      q_input = Q.Get<Tensor>().Data<T>();
+      k_input = q_input + num_heads_ * sequence_length * head_size;
+      q_rotary = RotaryQKV.GetMutable<Tensor>()->MutableData<T>();
+      k_rotary = q_rotary + num_heads_ * sequence_length * head_size;
+      Q = RotaryQKV;
+    } else {
+      OrtValue RotaryQ;
+      element_type = DataTypeImpl::GetType<T>();
+      std::vector<int64_t> q_dim({batch_size, num_heads_, sequence_length, head_size});
+      gsl::span<const int64_t> q_dim_span{q_dim};
+      TensorShape q_shape(q_dim_span);
+      Tensor::InitOrtValue(element_type, q_shape, allocator, RotaryQ);
+      OrtValue RotaryK;
+      std::vector<int64_t> k_dim({batch_size, kv_num_heads_, sequence_length, head_size});
+      gsl::span<const int64_t> k_dim_span{k_dim};
+      TensorShape k_shape(k_dim_span);
+      Tensor::InitOrtValue(element_type, k_shape, allocator, RotaryK);
+      q_input = Q.Get<Tensor>().Data<T>();
+      k_input = K.Get<Tensor>().Data<T>();
+      q_rotary = RotaryQ.GetMutable<Tensor>()->MutableData<T>();
+      k_rotary = RotaryK.GetMutable<Tensor>()->MutableData<T>();
+      Q = RotaryQ;
+      K = RotaryK;
+    }
+    ORT_RETURN_IF_ERROR(RunRotaryEmbedding<T>(tp, rotary_params, q_input,
+                                              pos_ids.Get<Tensor>().Data<int64_t>(), cos_cache->Data<T>(),
+                                              sin_cache->Data<T>(), q_rotary, rotary_interleaved_));
+    rotary_params.num_heads = kv_num_heads_;
+    rotary_params.hidden_size = parameters.kv_hidden_size;
+    if (!packed_qkv) {
+      rotary_params.batch_stride = kv_num_heads_ * rotary_params.head_stride;
+    }
+    ORT_RETURN_IF_ERROR(RunRotaryEmbedding<T>(tp, rotary_params, k_input,
+                                              pos_ids.Get<Tensor>().Data<int64_t>(), cos_cache->Data<T>(),
+                                              sin_cache->Data<T>(), k_rotary, rotary_interleaved_));
+    if (packed_qkv) {
```

Comment:
>packed_qkv [](http://example.com/codeflow?start=8&length=10)

if qkv packed, can we run the rotary_embedding at once:
for q, k, we do the rotary and for v we just copy.

Reply:
we could make changes to the rotary_embedding op to make this possible or copy paste the rotary code into a gqa helepr file, but not sure if this is a good idea?